backpack-compose/src/main/kotlin/net/skyscanner/backpack/compose/carousel/internal/BpkCarouselImpl.kt

```diff
@@ -0,0 +1,90 @@
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package net.skyscanner.backpack.compose.carousel.internal
+
+import androidx.compose.foundation.layout.Box
+import androidx.compose.foundation.layout.BoxScope
+import androidx.compose.runtime.Composable
+import androidx.compose.runtime.LaunchedEffect
+import androidx.compose.runtime.mutableStateOf
+import androidx.compose.runtime.saveable.rememberSaveable
+import androidx.compose.runtime.snapshotFlow
+import androidx.compose.ui.Alignment
+import androidx.compose.ui.Modifier
+import androidx.compose.ui.platform.testTag
+import com.google.accompanist.pager.ExperimentalPagerApi
+import com.google.accompanist.pager.HorizontalPager
+import com.google.accompanist.pager.rememberPagerState
+import kotlinx.coroutines.flow.distinctUntilChanged
+import net.skyscanner.backpack.compose.pageindicator.BpkPageIndicator
+import net.skyscanner.backpack.compose.pageindicator.BpkPageIndicatorStyle
+
+@OptIn(ExperimentalPagerApi::class)
+@Composable
+internal fun BpkCarouselImpl(
+  imageContent: @Composable (BoxScope.(Int) -> Unit),
+  count: Int,
+  currentImage: Int,
+  modifier: Modifier = Modifier,
+  onImageChanged: ((Int) -> Unit)? = null,
+) {
+  Box(modifier = modifier) {
+    val initialPageIndex = (Int.MAX_VALUE / 2) + currentImage
+    val (selectedPage, setSelectedPage) = rememberSaveable { mutableStateOf(currentImage) }
+
+    val pagerState = rememberPagerState(initialPageIndex).also {
+      LaunchedEffect(it) {
+        snapshotFlow { it.currentPage }.distinctUntilChanged().collect { index ->
+          val page = getModNumber(index, count)
+          setSelectedPage(page)
+          onImageChanged?.invoke(page)
+        }
+      }
+    }
+
+    HorizontalPager(
+      modifier = Modifier.testTag("pager"),
+      count = if (count > 1) Int.MAX_VALUE else 1, // if count > 1, set to Int.MAX_VALUE for infinite looping
+      state = pagerState,
+    ) { index ->
+      val page = getModNumber(index, count)
+      imageContent(page)
+    }
+
+    // if there is more than one image, display the page indicator
+    if (count > 1) {
+      BpkPageIndicator(
+        modifier = Modifier
+          .align(Alignment.BottomCenter)
+          .testTag("pageIndicator"),
+        totalIndicators = count,
+        currentIndex = selectedPage,
+        style = BpkPageIndicatorStyle.OverImage,
+      )
+    }
+  }
+}
+
+private fun getModNumber(index: Int, count: Int) = (index - (Int.MAX_VALUE / 2)).floorMod(count)
```

Comment:
So what's the purpose of this function?

Reply:
It's from this sample from accompanist. It calculates with modulo operation methods to get the real index of the current image, after we add the passed index on top of a large number (Int.MAX_VALUE / 2) for infinite looping use.
https://github.com/google/accompanist/blob/main/sample/src/main/java/com/google/accompanist/sample/pager/HorizontalPagerLoopingSample.kt